### PR TITLE
ft: S3C-4337 find compactable ranges

### PR DESCRIPTION
**Pull request by vrancurel**
Very simple algorithm to find ranges that are potentially compactable.

### CODE REVIEW

File: ldbreclaim/ldbreclaim.go [hunk 1]
```diff
@@ -0,0 +1,287 @@
...
+		ldbs:    s,
+		by:      by, // The Sort method's receiver is the function (closure) that defines the sort order.
+	}
+	sort.Sort(ps)
+}
+
+// ldbSorter joins a By function and a slice of Ldbs to be sorted.
+type ldbSorter struct {
+	ldbs []Ldb
+	by      func(p1, p2 *Ldb) bool // Closure used in the Less method.
```
Comment (jonathan-gramain) [inline ldbreclaim/ldbreclaim.go:62]: could you declare `by` of type `By` instead?

File: ldbreclaim/ldbreclaim.go [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+			resultsFound++
+		}
+		if resultsFound >= *numResults {
+			break
+		}
+	}
+}
+
+func Kind(k []byte) uint8 {
+        return k[len(k)-8]
```
Comment (jonathan-gramain) [inline ldbreclaim/ldbreclaim.go:190]: Could be interesting to have hints on the logic at play here, like a docstring for the `Kind` function and what is placed at `len(k)-8` or the general format of a key

**Review by jonathan-gramain — COMMENTED** [review]
I'm not sure why dumping the ldbs with the max number of keys first will give good ranges to compact, otherwise, the method adding the number of keys and deleted keys for overlapping ranges looks legit to me.

Could a different final sorting order give better results? I don't know, maybe a ratio of deleted/total keys, or combined with the absolute number of keys to avoid dumping ranges too small, but still taking into account the number of deleted keys in that range.

**Comment by vrancurel** [comment]
> Could a different final sorting order give better results? I don't know, maybe a ratio of deleted/total keys, or combined with the absolute number of keys to avoid dumping ranges too small, but still taking into account the number of deleted keys in that range.

If you just specify -t 500 -r 300 (default settings) it will display the ranges having worst case 500MB disk space and reclaim best case 300MB so 60% efficacy ratio.

**Comment by vrancurel** [comment]
@cmrh asks for the output of the tool to be easily plugged into the `curl` route.

**Comment by jonathan-gramain** [comment]
For reference, the compaction admin API: https://github.com/scality/MetaData/blob/development/7.10/documentation/PROTOCOL.ADMIN.md#compact-bucket-key-range

**Comment by vrancurel** [comment]
> > Could a different final sorting order give better results? I don't know, maybe a ratio of deleted/total keys, or combined with the absolute number of keys to avoid dumping ranges too small, but still taking into account the number of deleted keys in that range.
> 
> If you just specify -t 500 -r 300 (default settings) it will display the ranges having worst case 500MB disk space and reclaim best case 300MB so 60% efficacy ratio.

Can you suggest a comparison function ? is the ordering total ?

**Review by cmrh — COMMENTED** [review]
In an sample result as follows
`{001107.ldb 485027 compact-range/prefix-0346/data-file-0740.img compact-range/prefix-0898/data-file-0484.img 860474 308243 0}
{001109.ldb 484868 compact-range/prefix-0402/data-file-0774.img compact-range/prefix-0954/data-file-0543.img 796233 252451 0}
{001111.ldb 381957 compact-range/prefix-0458/data-file-0533.img compact-range/prefix-1000/data-file-1000.img 736238 196656 0}
{001104.ldb 1234225 compact-range/prefix-0499/data-file-0703.img compact-range/prefix-0800/data-file-1000.img 432878 131983 0}
`
You can either compact the data with the leveldb prefix as follows

`curl --header 'Content-Type: application/json' --request POST --data '{"prefix": "prefix-0868"}'  127.0.0.1:9146/_/buckets/compact-range/compactRange`
or alternatively use the startKey and endKey

`curl -v --header 'Content-Type: application/json' --request POST --data '{"startKey":"prefix-0458/data-file-0533.img","endKey":"prefix-1000/data-file-1000.img"}'  127.0.0.1:9146/_/buckets/compact-range/compactRange`
Note you cannot use the bucketName in the startKey/endKey

**Comment by vrancurel** [comment]
Closing this PR and re-creating it in the Metadata repository as it must be ran from within the MD container.

**Pull request closed without merging**